definitions/npm/hoist-non-react-statics_v3.x.x/flow_v0.103.x-/hoist-non-react-statics_v3.x.x.js

```diff
@@ -0,0 +1,84 @@
+declare module 'hoist-non-react-statics' {
+  /**
+   * Inspired by DefinitelyTyped/types/hoist-non-react-statics/index.d.ts
+   *
+   * Unfortunately, unlike in TypeScript, current flow definitions for React does not allow us to tell whether
+   * a React$ComponentType is “clean”, or had been the result of a call to React.memo or React.forwardRef.
+   * Therefore we’ll only be able to precisely filter out statics that are common to all 3 cases, and will
+   * have to live with maybes for everything else. This is not 100% precise, but is better than a blanket
+   * $Shape call on the source component’s statics like what we were doing before in the v2.x.x defs.
+   */
+
+  // declare type REACT_STATICS = {
+  //   childContextTypes: any,
+  //   contextType: any,
+  //   contextTypes: any,
+  //   defaultProps: any,
+  //   displayName: any,
+  //   getDefaultProps: any,
+  //   getDerivedStateFromError: any,
+  //   getDerivedStateFromProps: any,
+  //   mixins: any,
+  //   propTypes: any,
+  //   type: any,
+  //   ...
+  // };
+
+  // declare type MEMO_STATICS = {
+  //   $$typeof: any,
+  //   compare: any,
+  //   defaultProps: any,
+  //   displayName: any,
+  //   propTypes: any,
+  //   type: any,
+  //   ...
+  // };
+
+  // declare type FORWARD_REF_STATICS = {
+  //   $$typeof: any,
+  //   render: any,
+  //   defaultProps: any,
+  //   displayName: any,
+  //   propTypes: any,
+  //   ...
+  // };
+
+  declare type REACT_STATICS = {
+    // “Maybe” React statics
+    $$typeof?: any,
+    childContextTypes?: any,
+    compare?: any,
+    contextType?: any,
+    contextTypes?: any,
+    getDefaultProps?: any,
+    getDerivedStateFromError?: any,
+    getDerivedStateFromProps?: any,
+    mixins?: any,
+    render?: any,
+    type?: any,
+    // Common React statics
+    defaultProps: any,
+    displayName: any,
+    propTypes: any,
+    ...
+  };
+
+  declare type $HoistedStatics<S, C> = $Call<
+    & (empty => $Diff<S, REACT_STATICS>)
+    & (any => $Diff<S, $ObjMap<C, any> & REACT_STATICS>),
+    C
+  >;
+
+  /*
+    TP - target component props
+    T - target component statics
+    S - source component statics
+  */
+  declare function hoistNonReactStatics<TP, T, S, C: { [key: string]: true, ... }>(
+    TargetComponent: React$ComponentType<TP> & T,
+    SourceComponent: React$ComponentType<any> & S,
```

Comment:
Instead of `S` here, `{...S, ...REACT_STATICS}`

Now you don't need to diff them out!

Reply:
Thanks for the suggestion! This syntax does make things a lot cleaner, but I'm having trouble getting it to work 😭: https://flow.org/try/#0JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoctAGxQGc64BBOJADxiQDsATR5dDAB0AYVyQu3GAB4A3gB95AXwB8cWeThw6MFDGBo4KOAF4CKMpu279hgEamCdy1oDmSGEwAUASnVwiGABXKC5mIRRSOCVyGKoILh04ACFHLzAcMDoALnVFJT8TNS4g6moKZIjHfAsKoQczfGcKcgB6VrgYAE8wJBTaNABramAksw12rTtclC4ugBo2jqUWye7e-vRh0fhxpancnShgLldFyZXKdb7kge2xvKtp62PTxeUWniQaFCI4TCCXAwwASrA4UEEAGUbAY6NJISovFYRLkBBgACRicAJKQAFR6SGksy6agAZP4hJTIfM4JShLctiMkkpFj5cpCWmgEkkmJDHOwYBCMNC9LCvEwfBQuYl4Mk+WYBUKYCLbHQvMlJZRJgB1CClHhg3oYToACyQdD6MAgRjKrCgOCgcD18BgZq6cB4CXw8GySOlSRQAEZciUQHYkI6zLyIhRo2hY5ChO5PL4pdz4CgAEwhoJhiOOOUx8iF+PFxPJ7yaoA

Reply:
hrm :\ that's disappointing. this may be due to an issue with polymorphism and type destructors. If it doesn't work, ignore me :)

Reply:
@jbrown215 Indeed 😞These extracted statics definitely don't behave like our daily object types: https://flow.org/try/#0JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoctAGxQGc64BBOJADxiQDsATR5dDAB0AYVyQu3GAB4A3gB95AXwB8cWeThw6MFDGBo4KOAF4CKMpu279hgEamCdy1oDmSGEwAUASnVwiGABXKC5mIRRSOCVyGPIeJBoUIjhMIK4MYAgw9hgoQQBlGwM6aQKVLysRAC5EYhgAEjFwbKkAFQBPMCRpFC4OtQAyOAKAGnIfWoKKKmydZgKARkdc-IwivRKvJh8ZpiWIin3FoTsjg7AcGAgYLqQhd09fGbQ5+BRF2q4gkDskKEcx0O5CBaHOJ0e3l2IIuVxudweHgAQs9KAkkik0hl9NlWBw1jANrY6AB1YAwAAWTDaAHdgDoyhUqrUBBgmuJWlwYJ1ur1+kMRuNJnAGgARYCYTBlUbqVQvN4LABMK3xhWKaFJ5KptPpMG20JhiuB+yNZ0NQkuN3h3URTwNry48xQiq+Pz+ALMJuNBSNYPNkNRXst11uNseKIN6NomPSmVxqzVmw1ZMpTC4Nwp-x1DPKlS0NTqgnZLUkXJ5PT6Azgw1kcCE9YK0SFU3ljvg+wAzCq8oniSmqenKVm6Tp9XsCh3vZOzZ2LXDQ-cA-aFSgO67fv9AROp0I-bOl+dJ8HrYvkaigA